src/coreclr/tools/dotnet-pgo/Program.cs

```diff
@@ -388,32 +391,30 @@ static MibcConfig ParseMibcConfigsAndMerge(TypeSystemContext tsc, params PEReade
             return firstCfg;
         }
 
-        static int InnerMergeMain(CommandLineOptions commandLineOptions)
+        private int InnerMergeMain()
         {
-            if (commandLineOptions.InputFilesToMerge.Count == 0)
+            string outputPath = Get(_command.OutputFilePath);
+            if (outputPath == null)
             {
-                PrintUsage(commandLineOptions, "--input must be specified");
+                PrintError("Output filename must be specified");
                 return -8;
             }
 
-            if (commandLineOptions.OutputFileName == null)
-            {
-                PrintUsage(commandLineOptions, "--output must be specified");
-                return -8;
-            }
-
-            PEReader[] mibcReaders = new PEReader[commandLineOptions.InputFilesToMerge.Count];
+            var paths = Get(_command.InputFilesToMerge);
+            PEReader[] mibcReaders = new PEReader[paths.Count];
             for (int i = 0; i < mibcReaders.Length; i++)
             {
-                PrintMessage($"Opening {commandLineOptions.InputFilesToMerge[i].FullName}");
-                mibcReaders[i] = MIbcProfileParser.OpenMibcAsPEReader(commandLineOptions.InputFilesToMerge[i].FullName);
+                string path = paths.ElementAt(i).Value;
+                PrintMessage($"Opening {path}");
+                mibcReaders[i] = MIbcProfileParser.OpenMibcAsPEReader(path);
```

Comment:
The switch to `Dictionary` for references and the input .mibc files is concerning to me. It introduces non-determinism into the tool since these are enumerated. It seems like we should have a variant that expands into a `List` instead.

The fact that ilc/crossgen2 have the same problem for `--input-file-path` is a little concerning to me too, but I can see that it is preexisting to your changes.

Reply:
@jakobbotsch, in the existing implementation, we are using the same `AppendExpandedPaths` helper in: https://github.com/dotnet/runtime/blob/fb5f07f9c580bdcd5d0726d1391c2a52a01030f8/src/coreclr/tools/dotnet-pgo/CommandLineOptions.cs#L354

although it returns is a list, the result is equally non-deterministic (across multiple executions with same 'path patterns' arguments), due to the usage of `Dictionary` internally.

Note that we are (and were) using the same helper in ilc, crossgen2, ilverify and pgo.

Reply:
I wouldn't call it equally non-deterministic. In the common case without patterns the dictionaries have just one entry for each specified path and thus the existing version ends up being deterministic. This is for instance the case in our internal uses of `dotnet-pgo merge` in our training scenarios. There we are calling `dotnet-pgo merge` with a bunch of exact file names. The order was deterministic before but is going to be non-deterministic after this change.

Reply:
`AppendExpandedPaths` always uses `dictionary.Add(simpleName, fullFileName)` even for the exact file names.

Reply:
Yes, but `DefineFileOptionList` is creating a new dictionary for each path. Each dictionary ends up having only one entry and the list returned ends with the same order as the order given in the input.

Reply:
You are right. In ilc and crossgen2, we were passing the same dictionary instance to the helper before https://github.com/dotnet/runtime/blob/9b9b0c95c678065044d01049c56915e6b750b10a/src/coreclr/tools/aot/crossgen2/Program.cs#L141-L142 so the new (`BuildPathDictionary`) helper didn't change the behavior. Pgo is creating a new dictionary for each pattern, which is better.

BTW, this globbing implementation does not support patterns like `/path/**/*.mbic`, given: `/path/1/foo.mbic` and `/path/2/foo.mbic`, it will only give us `/path/2/foo.mbic`. The better approach would be to use something like [`MSBuildGlob`](https://github.com/dotnet/msbuild/blob/7c4597ea0ba65a2702f9f93016de94999c2a3c2c/src/Build/Globbing/MSBuildGlob.cs) from [this package](https://www.nuget.org/packages/Microsoft.Build).